rust/phoenix-channel/src/lib.rs

```diff
@@ -58,15 +62,78 @@ enum State {
     Closed,
 }
 
+async fn connect_websocket(
+    request: Request,
+    socket: tokio::net::TcpStream,
+) -> Result<
+    (WebSocketStream<MaybeTlsStream<TcpStream>>, Response),
+    tokio_tungstenite::tungstenite::Error,
+> {
+    client_async_tls(request, socket).await
+}
+
+async fn make_socket(
+    url: &Url,
+    socket_factory: &dyn SocketFactory<tokio::net::TcpSocket>,
+) -> Result<TcpStream, InternalError> {
+    let port = url
+        .port_or_known_default()
+        .expect("scheme to be http, https, ws or wss");
+    let addrs: Vec<SocketAddr> = match url.host().ok_or(InternalError::InvalidUrl)? {
+        Host::Domain(n) => tokio::net::lookup_host((n, port))
+            .await
+            .map_err(|_| InternalError::InvalidUrl)?
+            .collect(),
+        Host::Ipv6(ip) => {
+            vec![(ip, port).into()]
+        }
+        Host::Ipv4(ip) => {
+            vec![(ip, port).into()]
+        }
+    };
+
+    let mut last_error = None;
+    for addr in addrs {
+        let Ok(socket) = socket_factory(&addr) else {
+            continue;
+        };
+
+        match socket.connect(addr).await {
+            Ok(socket) => return Ok(socket),
+            Err(e) => {
+                last_error = Some(e);
+            }
+        }
+    }
+
+    let Some(err) = last_error else {
+        return Err(InternalError::InvalidUrl);
+    };
+
+    Err(InternalError::SocketConnection(err))
+}
+
+async fn create_and_connect_websocket(
+    url: Secret<LoginUrl>,
+    user_agent: String,
+    socket_factory: Arc<dyn SocketFactory<tokio::net::TcpSocket>>,
+) -> Result<WebSocketStream<MaybeTlsStream<TcpStream>>, InternalError> {
+    let socket = make_socket(url.expose_secret().inner(), &*socket_factory).await?;
+
+    let (stream, _) = connect_websocket(make_request(url, user_agent), socket)
+        .await
+        .map_err(InternalError::WebSocket)?;
+
+    Ok(stream)
+}
```

Comment:
I'd prefer utilities to be defined below where they are used, so in this order:

- `State::connect`
- `create_and_connect_websocket`
- `make_socket`
- `connect_websocket`

That way, one can read a file from top to bottom and the code goes from abstract to concrete.